sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -390,6 +390,7 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
       case If(TrueLiteral, trueValue, _) => trueValue
       case If(FalseLiteral, _, falseValue) => falseValue
       case If(Literal(null, _), _, falseValue) => falseValue
+      case If(_, trueValue, falseValue) if trueValue.semanticEquals(falseValue) => trueValue
```

Comment:
This is not right. The condition must be `deterministic`

Reply:
Can you elaborate? 

For `trueValue.semanticEquals(falseValue)`,  it's guaranteed that both `trueValue` and `falseValue` are `deterministic`.

```scala
def semanticEquals(other: Expression): Boolean =
    deterministic && other.deterministic && canonicalized == other.canonicalized
```

Reply:
Understandable that the condition can be non-deterministic, but this doesn't change the result of `If`.

Reply:
The condition could have a side effect. For example, calling a stateful UDF.

Reply:
Good point.